module/Finna/src/Finna/Search/R2/AuthenticationListener.php

```diff
@@ -0,0 +1,158 @@
+
+use Finna\Service\R2Service;
+use Finna\Service\RemsService;
+use FinnaSearch\Backend\R2\Connector;
+
+use Laminas\EventManager\EventInterface;
+use Laminas\EventManager\SharedEventManagerInterface;
+
+use VuFindSearch\Backend\BackendInterface;
+
+/**
+ * Restricted Solr (R2) Search authentication listener.
+ *
+ * @category VuFind
+ * @package  Search
+ * @author   Samuli Sillanpää <samuli.sillanpaa@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Site
+ */
+class AuthenticationListener
+{
+    /**
+     * Backend.
+     *
+     * @var BackendInterface
+     */
+    protected $backend;
+
+    /**
+     * Authentication manager
+     *
+     * @var Manager
+     */
+    protected $authManager;
+
+    /**
+     * R2 service
+     *
+     * @var R2Service;
+     */
+    protected $R2Service;
+
+    /**
+     * Connector
+     *
+     * @var FinnaSearch\Backend\R2\Connector
+     */
+    protected $connector;
+
+    /**
+     * Rems Service
+     *
+     * @var RemsService
+     */
+    protected $rems;
+
+    /**
+     * Constructor.
+     *
+     * @param BackendInterface $backend     Search backend
+     * @param R2Service        $R2Service   R2 service
+     * @param Connector        $connector   Backend connector
+     * @param RemsService      $rems        REMS service
+     *
+     * @return void
+     */
+    public function __construct(
+        BackendInterface $backend,
+        R2Service $R2Service,
+        Connector $connector,
+        RemsService $rems
+    ) {
+        $this->backend = $backend;
+        $this->R2Service = $R2Service;
+        $this->connector = $connector;
+        $this->rems = $rems;
+    }
+
+    /**
+     * Attach listener to shared event manager.
+     *
+     * @param SharedEventManagerInterface $manager Shared event manager
+     *
+     * @return void
+     */
+    public function attach(
+        SharedEventManagerInterface $manager
+    ) {
+        $manager->attach('VuFind\Search', 'pre', [$this, 'onSearchPre']);
+    }
+
+    /**
+     * Add username as a search parameter if the user is authorized.
+     *
+     * @param EventInterface $event Event
+     *
+     * @return EventInterface
+     */
+    public function onSearchPre(EventInterface $event)
+    {
+        $backend = $event->getTarget();
+        if ($backend === $this->backend) {
+            $params = $event->getParam('params');
+            $context = $event->getParam('context');
+            $this->connector->setUsername(null);
+            // Attempt to retrieve restricted metadata from the backend
+            // when the following holds:
+            // 1. The user is authorized
+            // (properly authenticated & registered to REMS)
+            // 2. If the search context is retrieve or retrieveBatch,
+            // and restricted metadata was requested.
+            if (!in_array($context, ['retrieve', 'retrieveBatch'])
```

Comment:
Vertailu väärin päin? Myös kommentin kohdat 1 ja 2 voisi olla toisin päin, kun tuo 2 tehdään ensin.

Reply:
Kommentti oli väärin. Suojattua metadataa pyydetään retrieve/retrieveBatch konteksteissa vain kun sitä on erikseen pyydetty ($paramsin kautta).

Reply:
Eli suojattua metadataa pyydetään kaikissa muissa konteksteissa aina? Jos näin, niin ok.

Reply:
Kyllä. En tiedä voiko muita konteksteja erotella järkevästi (milloin tarvitaan suojattua dataa ja milloin ei). Esim. hakutuloksessa on eroa jos siinä käytetään suojattuja kenttiä. Käytännössä myös retrieve/retrieveBatch käyttää suojattua dataa kun käytäjällä on oikeudet, koska LoaderFactory asettaa silloin Loaderille R2Restricted -parametrin, mitä kautta se kulkeutuu tänne. Eli jos luvitus on kunnossa, koodissa käsitellään suojattua driveria. Parempi olisi jos suojattu driveri voitaisiin ottaa vain tarvittaessa käyttöön. Tätä varten pitäisi ainakin yliajaa kaikki toiminnot, joiden pitää päästä kiinni suojattuun dataan. Kokeilin tätä alustavasti mutta arvioin että menee liian monimutkaiseksi.

Reply:
Yksinkertaistin tätä vielä jättämällä kontekstin tarkistuksen pois. AuthenticationListener pyytää suojattua metadataa aina kun käyttäjän luvitus on kunnossa.